### PR TITLE
Created initial guide for HDFS operators 

### PR DESCRIPTION
**Pull request by iadi7ya**
Guide for HDFS operators.

Created initial guide for HDFS operators explaining the Sensor Classes for hdfs available in existing code.
---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).

### CODE REVIEW

**Review by dimberman — APPROVED** [review]
@iadi7ya please get tests to pass and then LGTM

File: docs/howto/operator/apache/hdfs.rst [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+
+Prerequisite
+------------
+
+To use operators, you must configure a :doc:`HDFS Connection <../../connection/hdfs>`.
+
+.. _howto/operator:HdfsFolderSensor:
+
+Waits for a non-empty directory
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
Comment (mik-laj) [inline docs/howto/operator/apache/hdfs.rst]:
```suggestion
Waits for a non-empty directory
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

File: docs/howto/operator/apache/hdfs.rst [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+
+The :class:`~airflow.providers.apache.hdfs.sensors.hdfs.HdfsFolderSensor` operator is used to check for a non-empty directory in HDFS.
+
+Use the ``filepath`` parameter to poke until the provided file is found.
+
+.. exampleinclude:: /../airflow/providers/apache/hdfs/example_dags/
```
Comment (mik-laj) [inline docs/howto/operator/apache/hdfs.rst]: Here you must pass the full path to the file, not the directory.

File: docs/howto/operator/apache/hdfs.rst [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+    :language: python
+    :dedent: 4
+    :start-after: [START howto_operator_hdfs_folder_sensor]
+    :end-before: [END howto_operator_hdfs_folder_sensor]
+
+
+.. _howto/operator:HdfsRegexSensor:
+
+Waits for matching files by matching on regex
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
Comment (mik-laj) [inline docs/howto/operator/apache/hdfs.rst]:
```suggestion
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

File: docs/howto/operator/apache/hdfs.rst [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+    :language: python
+    :dedent: 4
+    :start-after: [START howto_operator_hdfs_regex_sensor]
+    :end-before: [END howto_operator_hdfs_regex_sensor]
+    
+    
+.. _howto/operator:HdfsSensor:
+
+Waits for a file or folder to land in HDFS
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
Comment (mik-laj) [inline docs/howto/operator/apache/hdfs.rst]:
```suggestion
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

File: docs/howto/operator/apache/hdfs.rst [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+
+The :class:`~airflow.providers.apache.hdfs.sensors.hdfs.HdfsSensor` operator is used to check for a file or folder to land in HDFS.
+
+Use the ``filepath`` parameter to poke until the provided file is found.
+
+.. exampleinclude:: /../airflow/providers/apache/hdfs/example_dags/
```
Comment (mik-laj) [inline docs/howto/operator/apache/hdfs.rst]: Here you must pass the full path to the file, not the directory.

File: docs/howto/operator/apache/hdfs.rst (hunk 2 above)
Comment (iadi7ya) [reply docs/howto/operator/apache/hdfs.rst]: Hi @mik-laj We don't have any example code file for hdfs in that directory.

File: docs/howto/operator/apache/hdfs.rst (hunk 5 above)
Comment (iadi7ya) [reply docs/howto/operator/apache/hdfs.rst]: Hi @mik-laj We don't have any example code file for hdfs in that directory.

File: docs/howto/operator/apache/hdfs.rst [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+    :language: python
+    :dedent: 4
+    :start-after: [START howto_operator_hdfs_sensor]
+    :end-before: [END howto_operator_hdfs_sensor]
+    
+
+
```
Comment (mik-laj) [inline docs/howto/operator/apache/hdfs.rst]:
```suggestion

```

File: docs/howto/operator/apache/hdfs.rst (hunk 4 above)
Comment (iadi7ya) [reply docs/howto/operator/apache/hdfs.rst]: @mik-laj Thank you!

File: docs/howto/operator/apache/hdfs.rst [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+Use the ``regex`` parameter to poke until the provided record is found. 
+
+.. exampleinclude:: /../airflow/providers/apache/hdfs/example_dags/
+    :language: python
+    :dedent: 4
+    :start-after: [START howto_operator_hdfs_regex_sensor]
+    :end-before: [END howto_operator_hdfs_regex_sensor]
+    
+
```
Comment (mik-laj) [inline docs/howto/operator/apache/hdfs.rst]:
```suggestion


```
I removed trailing whitepsaces

File: docs/howto/operator/apache/hdfs.rst (hunk 2 above)
Comment (mik-laj) [reply docs/howto/operator/apache/hdfs.rst]: Can you add these examples? Without examples, the guides are not very useful.  The most important thing in the guides for operators is that they show example operator usage, so examples are required.

Comment (iadi7ya) [reply docs/howto/operator/apache/hdfs.rst:43]: Hi @mik-laj I want to do that but I don't much idea about that.

**Review by iadi7ya — COMMENTED** [review]
Hi @dimberman Tests have passed. Ready for Merge

**Comment by mik-laj** [comment]
@iadi7ya The examples in the guides are very helpful and even required. We expect that users will be able to run these examples immediately after reading the guide. Is there no sample use of this operator in tests? Maybe you will be able to create a suitable example DAG based on this. 

In the case of guides for Google, we have even greater requirements. For Google, each guide has its own example DAG, but also system tests that run this DAG to validate it. If in the guide it is necessary to create additional resources, e.g. GCS buckets, they are also created by automatic scripts. In the case of Apache, this is not necessary, but I would like to show you some context why it is important to me that an example DAG be added.I try to make sure that each guide is of good quality. Not everyone has to meet all the requirements, but there are some rules that we try to follow in the guides.

We don't have general rules for how-to guides, but most of the tips for Google can apply to other services too. You can read the recommendations for Google guides in this document:: 
[https://docs.google.com/document/d/1RLQiuADIX00uGh0Llfcvv8WyCcayD6Cwkx4MNnHvDOc/edit#heading=h.wchyosj84oht](Google Operators Docs Guide)

**Comment by iadi7ya** [comment]
> @iadi7ya The examples in the guides are very helpful and even required. We expect that users will be able to run these examples immediately after reading the guide. Is there no sample use of this operator in tests? Maybe you will be able to create a suitable example DAG based on this?
> 
> In the case of guides for Google, we have even greater requirements. For Google, each guide has its own example DAG, but also system tests that run this DAG to validate it. If in the guide it is necessary to create additional resources, e.g. GCS buckets, they are also created by automatic scripts. In the case of Apache, this is not necessary, but I would like to show you some context why it is important to me that an example DAG be added.I try to make sure that each guide is of good quality. Not everyone has to meet all the requirements, but there are some rules that we try to follow in the guides.
> 
> We don't have general rules for how-to guides, but most of the tips for Google can apply to other services too. You can read the recommendations for Google guides:
> [Google Operators Docs Guide](https://docs.google.com/document/d/1RLQiuADIX00uGh0Llfcvv8WyCcayD6Cwkx4MNnHvDOc/edit#heading=h.wchyosj84oht) and [Google developer documentation style guide](https://developers.google.com/style)

> @iadi7ya The examples in the guides are very helpful and even required. We expect that users will be able to run these examples immediately after reading the guide. Is there no sample use of this operator in tests? Maybe you will be able to create a suitable example DAG based on this?
> 
> In the case of guides for Google, we have even greater requirements. For Google, each guide has its own example DAG, but also system tests that run this DAG to validate it. If in the guide it is necessary to create additional resources, e.g. GCS buckets, they are also created by automatic scripts. In the case of Apache, this is not necessary, but I would like to show you some context why it is important to me that an example DAG be added.I try to make sure that each guide is of good quality. Not everyone has to meet all the requirements, but there are some rules that we try to follow in the guides.
> 
> We don't have general rules for how-to guides, but most of the tips for Google can apply to other services too. You can read the recommendations for Google guides:
> [Google Operators Docs Guide](https://docs.google.com/document/d/1RLQiuADIX00uGh0Llfcvv8WyCcayD6Cwkx4MNnHvDOc/edit#heading=h.wchyosj84oht) and [Google developer documentation style guide](https://developers.google.com/style)

Hi @mik-laj Currently I am working on creating a connection doc for this. Thanks for the review and the documentation links. I will go through them and will then try to create a example dag. Thanks!

**Comment by iadi7ya** [comment]
Thanks @ryw :)

**Comment by kaxil** [comment]
Can you please rebase your PR on latest Master since we have applied [Black](https://github.com/apache/airflow/commit/4e8f9cc8d02b29c325b8a5a76b4837671bdf5f68) and [PyUpgrade](https://github.com/apache/airflow/commit/8c42cf1b00c90f0d7f11b8a3a455381de8e003c5) on Master.

It will help if your squash your commits into single commit first so that there are less conflicts.

File: docs/howto/operator/apache/hdfs.rst [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+ .. Unless required by applicable law or agreed to in writing,
+    software distributed under the License is distributed on an
+    "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+    KIND, either express or implied.  See the License for the
+    specific language governing permissions and limitations
+    under the License.
+
+
+
+Apache Hadoop HDFS Operators
```
Comment (kaxil) [inline docs/howto/operator/apache/hdfs.rst]: Same as https://github.com/apache/airflow/pull/11213#discussion_r549909399

This needs to be moved to https://github.com/apache/airflow/tree/master/docs/apache-airflow-providers-apache-hdfs

**Comment by kaxil** [comment]
I have updated and moved the docs to correct place since I didn't get a response from the author, let's wait for the CI results

**Review by kaxil — APPROVED** [review]

**Pull request merged**